### PR TITLE
[SH4RE-49] Edit/Delete Listing

### PR DESCRIPTION
**Pull request by WasifButt**
Changes:
- In view listing, I check if the listing belongs to user. If it does, replace the message button with edit and delete
- Upon pressing edit, goes to the Create listing view but passes the listing details as an binding object. If the values in listing are not null, the create listing view populates the fields
- If create listing is set to edit mode, post and cancel button are replaced with update and delete
- delete button removes listing from firestore
- update button updates values and reloads on viewListingView

https://user-images.githubusercontent.com/19561395/226703513-506e745a-abca-41f2-8613-58d3d6832f8a.mov

### CODE REVIEW

**Comment by WasifButt** [comment]
will add delete functionaility to this tomorrow

File: SH4RE/SH4RE/Account/MyListingsView.swift
Comment (americobarros) [inline SH4RE/SH4RE/Account/MyListingsView.swift:1]: Screenshot 2023-03-22 at 10 35 55 AM
The listings are too close to the edge, this can be solved by adding `.padding()` to the outermost VStack

File: SH4RE/SH4RE/Account/MyListingsView.swift [hunk 1]
```diff
@@ -35,8 +42,11 @@ struct MyListingsView: View {
         }
         .navigationTitle("My Listings")
         .onAppear(){
-            // TODO: change this to fetch current user listings (not all)
             fetchUsersListings(uid: getCurrentUserUid(), completion: { listings in
+                if (listings.isEmpty) {
+                    self.listingsView.listings.removeAll()
```
Comment (americobarros) [inline SH4RE/SH4RE/Account/MyListingsView.swift:48]: wait why do you need to remove all the listings from the listing view? Should it not already be empty?

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift [hunk 2]
```diff
@@ -210,57 +220,112 @@ struct CreateListingView: View {
...
-            //reset inputs
-            resetInputs()
-        }
+        return true
     }
     
     var body: some View {
         ZStack {
             Color.backgroundGrey.ignoresSafeArea()
+                .onAppear() {
```
Comment (americobarros) [inline SH4RE/SH4RE/CreateListing/CreateListingView.swift]: this code shouldnt be in the view body ideally since that should not really contain any logic (or ideally as little as possible), it makes the file harder to read. If possible, move to a function and call it and the end of the ZStack rather than attaching it to the color background.

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift [hunk 3]
```diff
@@ -210,57 +220,112 @@ struct CreateListingView: View {
...
+                    isEditing = editListing.title != ""
+                    if (isEditing) {
+                        title = editListing.title
+                        description = editListing.description
+                        cost = editListing.price
+                        postalCode = editListing.address
+                        availabilityCalendar.selectedDates = editListing.availability
+                        imagesCount = editListing.imagepath.count
+                        categorySelection = editListing.category
+                        imagesCount = (imagesCount <  5) ? imagesCount + 1 : imagesCount
```
Comment (americobarros) [inline SH4RE/SH4RE/CreateListing/CreateListingView.swift]: what does this do?

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift [hunk 4]
```diff
@@ -289,23 +369,28 @@ struct CreateListingView: View {
...
-                            showCancelAlertX.toggle()
-                            withAnimation(.easeInOut(duration: 1)) {
-                                value.scrollTo(1)
+                            if (isEditing) {
+                                self.presentationMode.wrappedValue.dismiss()
+                            }
+                            else {
+                                resetInputs()
+                                showCancelAlertX.toggle()
+                                withAnimation(.easeInOut(duration: 1)) {
```
Comment (americobarros) [inline SH4RE/SH4RE/CreateListing/CreateListingView.swift:437]: cmon animation

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift [hunk 5]
```diff
@@ -333,7 +418,25 @@ struct CreateListingView: View {
                     .frame(width: screenSize.width * 0.9, height: 130)
                     .background(.white)
                     .cornerRadius(30)
-                    
+                }
+                PopUp(show: $errorInUpdate) {
+                    VStack {
+                        Text("ERROR: No listing values changed")
```
Comment (americobarros) [inline SH4RE/SH4RE/CreateListing/CreateListingView.swift]: is this an error? if nothing changes do we not just want to leave it as it was?

**Review by americobarros — CHANGES_REQUESTED** [review]
Of the code which I could understand (and read in time for class) I made comments. I still need to do some testing in the simulator but here are my comments based only on a read through of the code!!!

File: SH4RE/SH4RE/Account/MyListingsView.swift (hunk 1 above)
Comment (WasifButt) [reply SH4RE/SH4RE/Account/MyListingsView.swift:48]: honestly can't explain this one, I just noticed that when `listings` was empty the `self.listingsView.listings` still had the last item left over, so I have to explicitly remove all from the array rather than assign it an empty array. It must have something to do with how Swift handles emptying array

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift (hunk 3 above)
Comment (WasifButt) [reply SH4RE/SH4RE/CreateListing/CreateListingView.swift]: adjusts image count so we can display the add image button if there are less than 5 images

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift (hunk 5 above)
Comment (WasifButt) [reply SH4RE/SH4RE/CreateListing/CreateListingView.swift]: i'll just rephrase it to, "Please update the some value first"

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift (hunk 4 above)
Comment (WasifButt) [reply SH4RE/SH4RE/CreateListing/CreateListingView.swift:437]: what do you mean?

**Comment by americobarros** [comment]
Nothing just nice jobOn Mar 22, 2023, at 11:51 AM, Wasif Butt ***@***.***> wrote:﻿
@WasifButt commented on this pull request.



In SH4RE/SH4RE/CreateListing/CreateListingView.swift:
>                          Button(action: {
-                            resetInputs()
-                            showCancelAlertX.toggle()
-                            withAnimation(.easeInOut(duration: 1)) {
-                                value.scrollTo(1)
+                            if (isEditing) {
+                                self.presentationMode.wrappedValue.dismiss()
+                            }
+                            else {
+                                resetInputs()
+                                showCancelAlertX.toggle()
+                                withAnimation(.easeInOut(duration: 1)) {

what do you mean?

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift (hunk 5 above)
Comment (americobarros) [reply SH4RE/SH4RE/CreateListing/CreateListingView.swift]: Wait but why do we need to have any pop up at all, when does this show up?

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift (hunk 4 above)
Comment (americobarros) [reply SH4RE/SH4RE/CreateListing/CreateListingView.swift:437]: nothing just nice job

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift [hunk 6]
```diff
@@ -210,57 +220,112 @@ struct CreateListingView: View {
...
+                }
             
             if (currentUser.isGuest()) {
                 GuestView(tabSelection: $tabSelection).environmentObject(currentUser)
             }
             else {
                 ScrollViewReader { value in
                     ScrollView([.vertical]) {
-                        Text("New Post")
+                        Text(isEditing ? "Edit Listing" : "New Post")
```
Comment (americobarros) [inline SH4RE/SH4RE/CreateListing/CreateListingView.swift:391]: please make these titles have the same positioning and styling as the home and search pages. Messaging will need to be updated too (separately)

File: SH4RE/SH4RE/Account/MyListingsView.swift (hunk 1 above)
Comment (americobarros) [reply SH4RE/SH4RE/Account/MyListingsView.swift:48]: i think it might have to do with how listings are instantiated? using `var listings = [Listing]()`. Strange. @bbryanb333 any thoughts?

File: SH4RE/SH4RE/CreateListing/CreateListingView.swift (hunk 6 above)
Comment (WasifButt) [reply SH4RE/SH4RE/CreateListing/CreateListingView.swift:391]: can you be more specific, is there a button on the home and search page that has the same style?

Comment (WasifButt) [reply SH4RE/SH4RE/CreateListing/CreateListingView.swift:391]: updated

**Comment by hannahbrooks** [comment]
it says "listing posted" after update is clicked, should it be listing updated instead?

**Review by hannahbrooks — CHANGES_REQUESTED** [review]
just the pop up
